rust/datafusion/src/execution/context.rs

```diff
@@ -73,26 +74,42 @@ impl ExecutionContext {
         }
     }
 
-    pub fn register_datasource(&mut self, name: &str, ds: Rc<RefCell<DataSource>>) {
-        self.datasources.borrow_mut().insert(name.to_string(), ds);
+    pub fn register_csv(
+        &mut self,
+        name: &str,
+        filename: &str,
+        schema: &Schema,
+        has_header: bool,
+    ) {
+        self.datasources.borrow_mut().insert(
+            name.to_string(),
+            Rc::new(CsvProvider::new(filename, schema, has_header)),
+        );
     }
 
-    pub fn execute(&mut self, plan: &LogicalPlan) -> Result<Rc<RefCell<Relation>>> {
-        println!("Logical plan: {:?}", plan);
+    fn optimize(&self, plan: &LogicalPlan) -> Rc<LogicalPlan> {
+        let rule: Rc<RefCell<OptimizerRule>> =
+            Rc::new(RefCell::new(ProjectionPushDown::new()));
+        let mut borrowed_rule = rule.borrow_mut();
+        borrowed_rule.optimize(plan).unwrap()
+    }
 
+    pub fn execute(&mut self, plan: &LogicalPlan) -> Result<Rc<RefCell<Relation>>> {
```

Comment:
Why the result is a `Rc<RefCell<Relation>>`? can we just use `Relation`?

Reply:
We need interior mutability when iterating over the results (record batches) of the relation with the current ownership model. I've started wondering if we should try and use lifetimes instead but I'm not sure yet if it is possible, especially for concurrent queries against the same data.

Reply:
Also Relation size isn't know at compile time so it has to be in `Box` or `Rc`

Reply:
That happens outside this method? If we have a unique reference to `Relation` then you can do whatever you want right? It's stronger than `Rc<RefCell<`.

Reply:
I'm getting to a point where I need to learn more about the ownership aspects of Rust and review how we are managing this. This may affect the design of DataSourceProvider / Relation as well. I think I will need to address this in a separate PR though.

Reply:
OK. Sounds good.